### PR TITLE
Fix issue#109

### PR DESCRIPTION
**Pull request by gaetanHerve**
Integrated in MR #298  

**Regression: doesn't work anymore after having merged pfm2**
_I think the problem doesn't come from conflicts resolution when merging,  but from functional incompatibility between this PR and a recent other one._

**TODO:** go back to previous commit and check for potential conflicts with last pull requests merged into pfm2 branch. 

enhancement : booking invoices can be generated taking quantities into account
By default, even if quantity field is not displayed in booking edition form, quantity = 1
Shouldn't have any impact on cases when not using quantity parameter

How it works:
     For each reservation, calculate its price.
     2 general cases:
     - booked resource price depends on duration, then it costs Unit price * reservation duration (in hours)
     - booked resource price depends on a quantity of elements, then it costs Unit price * nb elements (quantity)

     Unit prices can depend on the period booked (day, night, week-end)
     
     Specific case:
     - if a reservation depending on quantity of elements is stradding 2 types of period (i.e. night and day),
     then a ratio is applied.
     example:
     for a reservation covering 2 night hours and 6 day hours => nightRatio = 0.25 and dayRatio = 0.75
     if nightPrice = 20€/element and dayPrice = 10€/element,
     totalPrice = nbElements * (nightPrice * nightRatio) + nbElements * (dayPrice * dayRatio)

### CODE REVIEW

File: Modules/booking/Controller/BookinginvoiceController.php [hunk 1]
```diff
@@ -363,20 +348,29 @@ protected function invoice($id_space, $beginPeriod, $endPeriod, $id_resp, $numbe
...
-                    $userTime["nb_hours_we"] += $resaDayNightWe["nb_hours_we"];
+                    
+                    /* without taking resource quantity into account
+                    * $userTime["nb_hours_day"] += $resaDayNightWe["nb_hours_day"];
+                    * $userTime["nb_hours_night"] += $resaDayNightWe["nb_hours_night"];
+                    * $userTime["nb_hours_we"] += $resaDayNightWe["nb_hours_we"];
+                    */
+
+                    // taking resource quantity into account
+                    $qte = $resaDayNightWe["res_quantity"];
```
Comment (osallou) [inline Modules/booking/Controller/BookinginvoiceController.php:370]: not sure quantities should be multiplied by duration, should be checked to understand how it is expected to work

**Comment by osallou** [comment]
waiting for customer review

**Comment by gaetanHerve** [comment]
Customer review ok, but need to resolve functional conflicts with pfm2 branch

**Comment by gaetanHerve** [comment]
For our information: testing this MR I had some problems linked to the fact that "is_invoicing_unit" column wasn't present in my bk_calquantities table. However I had a fresh db install and CoreInstall::upgrade_v1_v2 add the column (line 120):
```
$bkqte = new BkCalQuantities();
$bkqte->addColumn("bk_calquantities", "is_invoicing_unit", "int(1)", 0);
```

File: Framework/Email.php [hunk 2]
```diff
@@ -23,9 +23,9 @@ class Email extends Model {
      * @param string $content
      * @param string $sentCopyToFrom
      * @param array  list of CoreFiles to attach to email
-     * @param bool   set toAddress as Bcc:, defaults to true, else just set in To:
+     * @param bool   set toAdress as Bcc:, defaults to true, else just set in To:
      */
-    public function sendEmail($from, $fromName, $toAddress, $subject, $content, $sentCopyToFrom = false, $files = [], $bcc=true ) {        
+    public function sendEmail($from, $fromName, $toAdress, $subject, $content, $sentCopyToFrom = false, $files = [], $bcc=true ) {
```
Comment (osallou) [inline Framework/Email.php:28]: regressio toaddress => toadress

File: Framework/Router.php [hunk 3]
```diff
@@ -345,13 +345,8 @@ private function createController($urlInfo, Request $request) {
      *        	Thrown exception
      */
     private function manageError(Throwable $exception, $type = '') {
-        $sendToSentry = true;
```
Comment (osallou) [inline Framework/Router.php:348]: regression, removes pfmauth handling

File: Modules/booking/Controller/BookingquantitiesController.php [hunk 4]
```diff
@@ -63,91 +64,59 @@ public function indexAction($id_space) {
...
-                       $packs[$supName[$p]] = $vid;
-                   }
-                }
-                $modelSups->setCalQuantity($id_space,  $supID[$p], $supResource[$p], $supName[$p], $supMandatory[$p]);
-            }
+            // format into arrays
+            $supIsInvoicingUnit = is_array($supIsInvoicingUnit) ? $supIsInvoicingUnit : [$supIsInvoicingUnit];
+            $supID = is_array($supID) ? $supID : [$supID];
 
-            /* bug to get last id (could conflict)
```
Comment (osallou) [inline Modules/booking/Controller/BookingquantitiesController.php:103]: regression, remove commented code and removes above corevirtual use

File: Modules/booking/Controller/BookingschedulingController.php [hunk 5]
```diff
@@ -50,67 +50,56 @@ public function indexAction($id_space) {
...
-            $bkScheduling["id_rearea"] = $id_rearea;
-        }
 
         $form = new Form($this->request, "bookingschedulingedit");
         $form->setTitle(BookingTranslator::Edit_scheduling($lang) . ": " . $name, 3);
-        $form->addChoicesList(BookingTranslator::Availables_days($lang),
+        $form->addChoicesList(BookingTranslator::Availables_days($lang), 
                 BookingTranslator::DaysList($lang), 
                 array("is_monday", "is_tuesday", "is_wednesday", "is_thursday", "is_friday", "is_saturday", "is_sunday"),
-                array($bkScheduling["is_monday"], $bkScheduling["is_tuesday"], $bkScheduling["is_wednesday"], $bkScheduling["is_thursday"], $bkScheduling["is_friday"], $bkScheduling["is_saturday"], $bkScheduling["is_sunday"])
```
Comment (osallou) [inline Modules/booking/Controller/BookingschedulingController.php:78]: regression on naming

File: Modules/booking/Model/BkCalQuantities.php [hunk 6]
```diff
@@ -131,7 +132,7 @@ public function setCalQuantity($id_space, $id_quantity, $id_resource, $name, $ma
      * @return boolean
      */
     public function isCalQuantityId($id_space, $id_quantity, $id_resource) {
-        $sql = "select id from bk_calquantities where id_quantity=? AND id_resource=? AND deleted=0 AND id_space=?";
```
Comment (osallou) [inline Modules/booking/Model/BkCalQuantities.php:134]: removes deleted=0

File: Modules/booking/Model/BkScheduling.php [hunk 7]
```diff
@@ -40,7 +40,7 @@ public function getDefault() {
             "is_wednesday" => 1, "is_thursday" => 1, "is_friday" => 1,
             "is_saturday" => 1, "is_sunday" => 1, "day_begin" => 8,
             "day_end" => 18, "size_bloc_resa" => 3600, "booking_time_scale" => 1,
-            "resa_time_setting" => 1, "default_color_id" => 1, "id_rearea" => 0);
```
Comment (osallou) [inline Modules/booking/Model/BkScheduling.php:43]: removes rearea

**Review by osallou — CHANGES_REQUESTED** [review]

**Pull request closed without merging**